lib/phoenix/debug.ex

```diff
@@ -0,0 +1,88 @@
+defmodule Phoenix.Debug do
+  @moduledoc """
+  Functions for runtime introspection and debugging of Phoenix applications.
+
+  TODO
+  """
+
+  @doc """
+  Returns a list of all currently connected `Phoenix.Socket` transport processes.
+
+  Note that custom sockets implementing the `Phoenix.Socket.Transport` behaviour
+  are not listed.
+
+  Each process corresponds to one connection that can have multiple channels.
+
+  For example, when using Phoenix LiveView, the browser establishes a socket
+  connection when initially navigating to the page, and each live navigation
+  retains the same socket connection. Nested LiveViews also share the same
+  connection, each being a different channel. See `Phoenix.Debug.channels/1`.
+
+  ## Examples
+
+      iex> Phoenix.Debug.list_channel_sockets()
+      [#PID<0.123.0>]
+
+  """
+  def list_channel_sockets do
+    for pid <- Process.list(), channel_socket?(pid) do
+      pid
+    end
+  end
+
+  @doc """
+  Returns true if the given pid is a `Phoenix.Socket` transport process.
+
+  ## Examples
+
+      iex> Phoenix.Debug.list_channel_sockets() |> Enum.at(0) |> channel_socket?()
+      true
+
+      iex> channel_socket?(pid(0,456,0))
+      false
+
+  """
+  def channel_socket?(pid) do
+    # `Phoenix.Socket`s set the "$process_label" to {Phoenix.Socket, handler_module, id}
+    with info when is_list(info) <- Process.info(pid, [:dictionary]),
+         {:dictionary, dictionary} <- List.keyfind(info, :dictionary, 0),
+         {:"$process_label", label} <- List.keyfind(dictionary, :"$process_label", 0),
+         {Phoenix.Socket, mod, id} when is_atom(mod) and (is_binary(id) or is_nil(id)) <- label do
+      true
+    else
+      _ -> false
+    end
+  end
+
+  @doc """
+  Returns a list of all currently connected channels for the given `Phoenix.Socket` pid.
+
+  ## Examples
+
+      iex> Phoenix.Debug.list_channel_sockets() |> Enum.at(0) |> Phoenix.Debug.channels()
+      {:ok,
+       [
+         %{pid: #PID<0.1702.0>, status: :joined, topic: "lv:phx-GDp9a9UZPiTxcgnE"},
+         %{pid: #PID<0.1727.0>, status: :joined, topic: "lv:sidebar"}
+       ]}
+
+      iex> Phoenix.Debug.channels(pid(0,456,0))
+      {:error, :not_alive}
+
+  """
+  def channels(socket_pid) do
+    ref = make_ref()
```

Comment:
Should this perhaps be called `list_channels` and go through a similar code as `list_channels_sockets`, which uses `Process.info` to see if they are a channel or not, rather than sending a debug channels message? Especially if we are thinking about making as straight-forward to use as possible?

Reply:
I would kinda of expect the channels to not include LiveViews indeed but if you disagree, then I think this is good to go. Perhaps we just add `list_channels` as a shortcut (i.e. a `flat_map`).

Reply:
I kept the list functions "cheap" for now, returning a map but only with information that does not require sending a message to the pid.